File: kotlin/src/main/kotlin/io/ExcelReader.kt
package io

import models.Student
import org.apache.poi.ss.usermodel.CellType
import org.apache.poi.ss.usermodel.WorkbookFactory
import java.io.File

/**
 * Reads student data from an Excel (`.xlsx`) file using Apache POI.
 *
 * The expected spreadsheet layout is:
 *
 * | StudentID | Name | Subject1 | Subject2 | … |
 * |-----------|------|----------|----------|---|
 *
 * Row 0 is treated as the header row and is skipped when reading data.
 *
 * @property filePath Path to the `.xlsx` file to read.
 */
class ExcelReader(private val filePath: String) {

    /**
     * Reads the first sheet of the Excel file and returns a list of [Student]
     * objects populated with their ID, name, and raw scores.
     *
     * Computed fields (average, grade, GPA, status) are left at their defaults
     * and should be filled in by [services.GradeCalculator].
     */
    fun read(): List<Student> {
        // Step 1 — Open the workbook from the file path using Apache POI's factory
        val workbook = WorkbookFactory.create(File(filePath))

        // Step 2 — Grab the first (and usually only) sheet
        val sheet = workbook.getSheetAt(0)

        val students = mutableListOf<Student>()

        // Step 3 — Iterate over rows, skipping the header at index 0
        for (rowIndex in 1..sheet.lastRowNum) {
            val row = sheet.getRow(rowIndex) ?: continue // Skip null rows

            // Column 0 → Student ID (read as string)
            val studentId = getCellValueAsString(row.getCell(0))

            // Column 1 → Student name
            val name = getCellValueAsString(row.getCell(1))

            // Columns 2+ → Subject scores (read as doubles)
            val scores = mutableListOf<Double>()
            // lastCellNum is 1-based (returns the index of the last cell + 1)
            val lastCol = row.lastCellNum.toInt()
            for (col in 2 until lastCol) {
                val cell = row.getCell(col)
                if (cell != null) {
                    // getNumericCellValue() returns the numeric value of the cell
                    scores.add(cell.numericCellValue)
                }
            }

            // Step 4 — Create a Student with only the raw data
            students.add(
                Student(
                    studentId = studentId,
                    name = name,
                    scores = scores
                )
            )
        }

        // Step 5 — Close the workbook to release file resources
        workbook.close()

        return students
    }

    /**
     * Reads the header row and returns only the subject column names
     * (i.e. everything after StudentID and Name).
     *
     * This is useful for [io.ExcelWriter] so it can reproduce the same subject
     * headers in the output file.
     */
    fun readSubjectHeaders(): List<String> {
        val workbook = WorkbookFactory.create(File(filePath))
        val sheet = workbook.getSheetAt(0)

        // The header row is row 0
        val headerRow = sheet.getRow(0)
        val headers = mutableListOf<String>()

        // Extract column names starting from index 2 (skip StudentID, Name)
        val lastCol = headerRow.lastCellNum.toInt()
        for (col in 2 until lastCol) {
            val cell = headerRow.getCell(col)
            if (cell != null) {
                headers.add(cell.stringCellValue)
            }
        }

        workbook.close()
        return headers
    }

    /**
     * Helper to read a cell value as a [String], regardless of whether the
     * cell is stored as a numeric or string type in Excel.
     */
    private fun getCellValueAsString(cell: org.apache.poi.ss.usermodel.Cell?): String {
        if (cell == null) return ""
        return when (cell.cellType) {
            CellType.STRING  -> cell.stringCellValue
            CellType.NUMERIC -> cell.numericCellValue.toLong().toString()
            else             -> cell.toString()
        }
    }
}

File: dart/lib/io/excel_reader.dart
import 'dart:io';
import 'package:excel/excel.dart';

import '../models/student.dart';

/// Reads student data from an Excel (`.xlsx`) file.
///
/// The expected spreadsheet layout is:
///
/// | StudentID | Name | Subject1 | Subject2 | … |
/// |-----------|------|----------|----------|---|
///
/// Row 0 is treated as the header row and is skipped when reading data.
class ExcelReader {
  /// Path to the `.xlsx` file to read.
  final String filePath;

  /// Creates an [ExcelReader] that will read from [filePath].
  ExcelReader(this.filePath);

  /// Reads the first sheet of the Excel file and returns a list of [Student]
  /// objects populated with their ID, name, and raw scores.
  ///
  /// Computed fields (average, grade, GPA, status) are left at their defaults
  /// and should be filled in by [GradeCalculator].
  List<Student> read() {
    // Step 1 — Read the raw bytes of the file from disk
    final fileBytes = File(filePath).readAsBytesSync();

    // Step 2 — Decode the bytes into an Excel workbook object
    final excel = Excel.decodeBytes(fileBytes);

    // Step 3 — Grab the first (and usually only) sheet
    final sheetName = excel.tables.keys.first;
    final sheet = excel.tables[sheetName]!;

    // We'll collect all Student objects here
    final List<Student> students = [];

    // Step 4 — Iterate over the rows, skipping the header (index 0)
    for (var rowIndex = 1; rowIndex < sheet.maxRows; rowIndex++) {
      final row = sheet.rows[rowIndex];

      // Skip completely empty rows
      if (row.isEmpty || row[0] == null) continue;

      // Column 0 → Student ID (converted to String)
      final studentId = row[0]!.value.toString();

      // Column 1 → Student name
      final name = row[1]!.value.toString();

      // Columns 2+ → Subject scores (converted to double)
      final scores = <double>[];
      for (var col = 2; col < row.length; col++) {
        if (row[col] != null && row[col]!.value != null) {
          // Parse the cell value; it may be int or double in Excel
          scores.add(double.parse(row[col]!.value.toString()));
        }
      }

      // Step 5 — Create a Student with only the raw data
      students.add(Student(
        studentId: studentId,
        name: name,
        scores: scores,
      ));
    }

    return students;
  }

  /// Reads the header row and returns only the subject column names
  /// (i.e. everything after StudentID and Name).
  ///
  /// This is useful for [ExcelWriter] so it can reproduce the same subject
  /// headers in the output file.
  List<String> readSubjectHeaders() {
    // Read and decode the file (same as in read())
    final fileBytes = File(filePath).readAsBytesSync();
    final excel = Excel.decodeBytes(fileBytes);

    // Grab the first sheet
    final sheetName = excel.tables.keys.first;
    final sheet = excel.tables[sheetName]!;

    // The header row is row 0
    final headerRow = sheet.rows[0];

    // Extract column names starting from index 2 (skip StudentID, Name)
    final headers = <String>[];
    for (var col = 2; col < headerRow.length; col++) {
      if (headerRow[col] != null && headerRow[col]!.value != null) {
        headers.add(headerRow[col]!.value.toString());
      }
    }

    return headers;
  }
}
